tests/test-dladdr.c

```diff
@@ -0,0 +1,192 @@
+    {
+        passed = expected_result == NoInfo || expected_result == Fail;
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match && !info.dli_sname)
+                 || (expected_result == PassWithDifferentAddress && sym_match && !addr_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
+    }
+    printf( "check address %p which has symbol '%s' -> %s%s", addr, addrsym, passed ? "passed" : "failed", verbose || !passed ? " " : "\n" );
+    if( verbose || !passed )
+    {
+        if( !result )
+            printf( "(could not get symbol information for address %p)\n", addr );
+        else
+            print_dl_info( &info, "(", ")\n");
+    }
+    return !passed;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#include <windows.h>
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+/* hide warning "reclared without 'dllimport' attribute" */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#elif defined(_MSC_VER)
+/* disable warning C4273 inconsistent dll linkage */
+#pragma warning(push)
+#pragma warning(disable: 4273)
+#endif
+/*
+ * Windows API functions decorated with WINBASEAPI are imported from iat.
+ * For testing force linking by import thunk for the following functions
+ */
+HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
+/* force linking by iat */
+/* We can't use WINBASEAPI here because i586-mingw32msvc-gcc doesn't define it correctly */
```

Comment:
As written in the last update https://github.com/dlfcn-win32/dlfcn-win32/pull/72#discussion_r542229131 `WINBASEAPI` does not have to be set to the `__declspec(dllimport)`. It can be also to the other value. It just defines how windows base api functions should be linked and it can be overidden by compiler/application. So it is not problem of i586-mingw32msvc-gcc, it defines WINBASEAPI correctly.